jOOQ/src/main/java/org/jooq/tools/Convert.java

```diff
@@ -517,7 +522,18 @@ else if (Collection.class.isAssignableFrom(toClass) &&
                 else if (toClass.isArray()
                         && Collection.class.isAssignableFrom(fromClass)){
                     Collection f = (Collection) from;
-                    return (U) f.toArray();
+                    Class componentType = toClass.getComponentType();
+
+                    Object[] dest = (Object[]) Array.newInstance(componentType, f.size());
+                    Object[] list = f.stream()
+                            .map(e -> {
+                                if (!componentType.isAssignableFrom(e.getClass()))
+                                    return convert(e, componentType);
+                                return e;
+                            }).toArray();
+                    System.arraycopy(list, 0, dest, 0, dest.length);
+
+                    return (U) dest;
                 }
```

Comment:
I'm aware that Java 8 logic is scattered across this code (including references to `Optional`), but you don't see the full picture, unfortunately. The commercial distribution has special markers that delimit code with a Java 8 dependency and strips it from the Java 6 distribution. The Open Source edition doesn't have those markers.